### PR TITLE
[SPARK-49569][CONNECT][SQL] Add shims to support SparkContext and RDD

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR does two things:
- It adds shims for SparkContext and RDD. These are in a separate module. This module is a compile time dependency for sql/api, and a regular dependency for connector/connect/client/jvm. We remove this dependency in catalyst and connect-server because those should use the actual implementation.
- It adds RDD (and the one SparkContext) based method to the shared Scala API. For connect these methods throw an unsupported operation exception.

### Why are the changes needed?
We are creating a shared Scala interface for Classic and Connect.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existing tests. I will add a couple on the connect side.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/package.scala [hunk 1]
```diff
@@ -52,4 +52,7 @@ package object sql {
     f(builder)
     column(builder.build())
   }
+
+  private[sql] def throwRddNotSupportedException(): Nothing =
+    throw new UnsupportedOperationException("RDDs are not supported in Spark Connect.")
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/package.scala:57]: Make this an actual error.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 2]
```diff
@@ -84,10 +87,14 @@ class SparkSession private[sql] (
 
   private[sql] val observationRegistry = new ConcurrentHashMap[Long, Observation]()
 
-  private[sql] def hijackServerSideSessionIdForTesting(suffix: String) = {
+  private[sql] def hijackServerSideSessionIdForTesting(suffix: String): Unit = {
     client.hijackServerSideSessionIdForTesting(suffix)
   }
 
+  /** @inheritdoc */
+  override def sparkContext: SparkContext =
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:95]: Make this use the error framework.

**Comment by hvanhovell** [comment]
cc @grundprinzip @HyukjinKwon

**Review by HyukjinKwon — APPROVED** [review]

**Review by dongjoon-hyun — COMMENTED** [review]
Could you make CIs happy, @hvanhovell ?
```
[error] /home/runner/work/spark/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:83:18: value aggregate is not a member of org.apache.spark.rdd.RDD[Array[String]]
[error]         tokenRDD.aggregate(startType)(inferRowType, mergeRowTypes)
```

**Comment by hvanhovell** [comment]
@dongjoon-hyun yeah working on it. SBT does not seem to respect maven exclusions...

**Comment by dongjoon-hyun** [comment]
Could you rebase once more, @hvanhovell ?

**Comment by hvanhovell** [comment]
@dongjoon-hyun it is a bit more complicated. While SBT compile seems to respect Maven exclusions, test:compile and package do not. Investigating. I'd personally not hold off preview2 because of this.

**Comment by dongjoon-hyun** [comment]
Ack! Thank you for sharing the status, @hvanhovell .

**Comment by dongjoon-hyun** [comment]
Could you re-trigger the failed CIs?
```
[info] ClientE2ETestSuite:
[info] org.apache.spark.sql.ClientE2ETestSuite *** ABORTED *** (2 minutes, 48 seconds)
[info]   The code passed to eventually never returned normally. Attempted 1 times over 2.805218503466667 minutes. (RemoteSparkSession.scala:197)
[info]   org.scalatest.exceptions.TestFailedDueToTimeoutException:
```

**Comment by hvanhovell** [comment]
Merging to master.

**Comment by LuciferYang** [comment]
This PR has caused the Maven daily test build to fail: xxx. 

- https://github.com/apache/spark/actions/runs/11255598249
- https://github.com/apache/spark/actions/runs/11256610976

```
scaladoc error: fatal error: object scala in compiler mirror not found.
Error:  Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.9.1:doc-jar (attach-scaladocs) on project spark-connect-shims_2.13: MavenReportException: Error while creating archive: wrap: Process exited with an error: 1 (Exit value: 1) -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Error:  
Error:  After correcting the problems, you can resume the build with the command
Error:    mvn <args> -rf :spark-connect-shims_2.13
Error: Process completed with exit code 1.
```

I am trying to fix it at: https://github.com/apache/spark/pull/48399

**Pull request closed without merging**